### PR TITLE
RA-1603 : <encounterDiagnoses*/> to specify a diagnosesSet to search …

### PR DESCRIPTION
**Pull request by Ruhanga**
Ticket : https://issues.openmrs.org/browse/RA-1603

Missing tests. I will add tests in the next commit.

### CODE REVIEW

File: api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java [hunk 1]
```diff
@@ -61,6 +61,8 @@
     private UiUtils uiUtils;
     private String selectedDiagnosesTarget;
 
+    private String diagnosisSets;
```
Comment (mogoodrich) [inline api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java]: I'd recommend naming this something like 'diagnosisSetUuids" to make it clearer that this is still just a string of uuid and not a Collection<Concept>.  (Ideally this *would* be a Collection<Concept> but I realized that wouldn't make a ton of sense in this case, since you have to convert it right back to a string afterwards.

File: api-pre2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java [hunk 2]
```diff
@@ -65,6 +65,8 @@
 
     private ConceptService conceptService;
 
+    private String diagnosisSets;
```
Comment (mogoodrich) [inline api-pre2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java]: See comment above.

File: api/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesByObsElement.java [hunk 3]
```diff
@@ -66,6 +66,8 @@
 
     private ConceptService conceptService;
 
+    private String diagnosisSets;
```
Comment (mogoodrich) [inline api/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesByObsElement.java]: See comment above.

File: omod/src/main/webapp/resources/scripts/diagnoses/diagnoses-angular.js [hunk 4]
```diff
@@ -4,7 +4,7 @@ var app = angular.module('diagnoses', [])
             // I don't know how to use an angular template programmatically, so use an underscore template instead. :-(
             var itemFormatter = _.template($('#' + attrs.itemformatter).html());
             element.autocomplete({
-                source: emr.fragmentActionLink("coreapps", "diagnoses", "search"),
+                source: emr.fragmentActionLink("coreapps", "diagnoses", "search", {diagnosisSets: attrs.diagnosissets}),
```
Comment (mogoodrich) [inline omod/src/main/webapp/resources/scripts/diagnoses/diagnoses-angular.js]: When you test, just make sure you test that when diagnosisSets is null, this still operates correctly (ie doesn't break existing functionality).

File: api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java (hunk 1 above)
Comment (Ruhanga) [reply api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java]: Thank you @mogoodrich. Yes, this makes sense and perhaps, I'd simply do `diagnosisSetIds` to remain consistent with the way attributes in HFE tags are named.

File: omod/src/main/webapp/resources/scripts/diagnoses/diagnoses-angular.js (hunk 4 above)
Comment (Ruhanga) [reply omod/src/main/webapp/resources/scripts/diagnoses/diagnoses-angular.js]: Ok, this is handled at [form creation](https://github.com/openmrs/openmrs-module-coreapps/pull/199/files#diff-e0e5822659198b9ceada75532b7be803R78) for the form and [back-end here](https://github.com/openmrs/openmrs-module-coreapps/pull/198/commits/6772566d8abb1378b01197f0d9d4f5bd14e1d7b8#diff-15549579e8c094656a9104bb0e207449R59) .

Comment (mogoodrich) [reply omod/src/main/webapp/resources/scripts/diagnoses/diagnoses-angular.js]: Cool thanks @Ruhanga, just wanted to make sure you tested...

File: api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java (hunk 1 above)
Comment (mks-d) [reply api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java]: @mogoodrich, it would be _very_ nice is concept mappings could be used to specify the sets to work with. I find UUIDs very impractical and difficult to read in HFE forms.

Would that be ok with you?

The tag handler would parse the list and make sure that only UUIDs are sent over to the underlying web service (`DiagnosesFragmentController`).

Comment (mogoodrich) [reply api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java]: oh, absolutely @mks-d  that would be the ideal... then this *could* become a List<Concept>... the attribute on the encounterDiagnosis tag could be diagnosisSets, and you would use this existing method to parse it, as it supports our multiple methods of referencing concepts (including using mappings):

https://github.com/openmrs/openmrs-module-htmlformentry/blob/master/api/src/main/java/org/openmrs/module/htmlformentry/HtmlFormEntryUtil.java#L534

Comment (mks-d) [reply api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java]: @mogoodrich the above link is pointing in between two methods, but I'm sure @Ruhanga will figure it out ;-)

Comment (Ruhanga) [reply api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java]: Yes, thank you @mogoodrich.

Comment (Ruhanga) [reply api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java]: I have just addressed this is the commits https://github.com/openmrs/openmrs-module-coreapps/pull/199/commits/d979720b4b7b6d2b732728289d6b6c12c32ffe92 and https://github.com/openmrs/openmrs-module-coreapps/pull/199/commits/ca271c8eb14beff9177e00397f84aca2d747f6a0 .

Comment (Ruhanga) [reply api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java]: Sorry about the CI failure. It is a log length issue that prematurely terminates the compilation.

File: api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java [hunk 5]
```diff
@@ -158,6 +164,39 @@ public String generateHtml(FormEntryContext context) {
...
     }
+    
+    /**
+     * This method recieves a comma seperated list of diagnosis sets which are verified and then
+     * returns as comma seperated list of their uuids. diagnosisSetIds may be concept's 1). uuid
+     * 2). mapping, 3). id
+     *
+     * @param diagnosisSetIds
+     * @return uuids
+     */
```
Comment (mks-d) [inline api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java]:
```java
/**
 * This method receives a comma-separated list of diagnosis sets 'identifiers'
 * and returns as comma-separated list of their uuids.
 *
 * @param diagnosisSetIds Either concepts UUIDS, mappings or internal IDs.
 * @return The list of diagnoses sets as a list of their concept UUIDs.
 * @throws IllegalArgumentException if one or more sets cannot be fetched from the database.
 */
```

File: api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java [hunk 6]
```diff
@@ -158,6 +164,39 @@ public String generateHtml(FormEntryContext context) {
...
+        if ("".equals(diagnosisSetIds)) {
+            return diagnosisSetIds;
+        }
+        List<Concept> concepts = new ArrayList<Concept>();
+        for (StringTokenizer st = new StringTokenizer(diagnosisSetIds, ","); st.hasMoreTokens();) {
+            String id = st.nextToken().trim();
+            Concept concept = HtmlFormEntryUtil.getConcept(id);
+            if (concept == null) {
+                throw new IllegalArgumentException("Cannot find diagnosis set for value " + id
+                        + " in diagnosisSets attribute value. Parameters: " + diagnosisSetIds);
```
Comment (mks-d) [inline api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java]:
```java
IllegalArgumentException("Cannot fetch the diagnosis set identified by '" + id
                        + "'' in the provided list of diagnoses sets: " + diagnosisSetIds);
```

File: api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java [hunk 7]
```diff
@@ -158,6 +164,39 @@ public String generateHtml(FormEntryContext context) {
...
+        }
+        StringBuilder sb = new StringBuilder("");
+        if (CollectionUtils.isNotEmpty(concepts)) {
+            for (Concept con : concepts) {
+                sb.append(con.getUuid() + ",");
+            }
+        }
+        String uuids = sb.toString();
+
+        return uuids.substring(0, uuids.lastIndexOf(","));
```
Comment (mks-d) [inline api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java]: To make this a little safer, have a look at this: https://stackoverflow.com/a/15900408/321797

File: api-pre2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java [hunk 8]
```diff
@@ -136,6 +142,39 @@ public String generateHtml(FormEntryContext context) {
...
+    
+    /**
+     * This method recieves a comma seperated list of diagnosis sets which are verified and then
+     * returns as comma seperated list of their uuids. diagnosisSetIds may be concept's 1). uuid
+     * 2). mapping, 3). id
+     *
+     * @param diagnosisSetIds
+     * @return uuids
+     */
+    private String validateAndFormat(String diagnosisSetIds) {
```
Comment (mks-d) [inline api-pre2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java]: Why is this here again?

File: api/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesByObsElement.java [hunk 9]
```diff
@@ -138,6 +144,39 @@ public String generateHtml(FormEntryContext context) {
...
 
+    /**
+     * This method recieves a comma seperated list of diagnosis sets which are verified and then
+     * returns as comma seperated list of their uuids. diagnosisSetIds may be concept's 1). uuid
+     * 2). mapping, 3). id
+     *
+     * @param diagnosisSetIds
+     * @return uuids
+     */
+    private String validateAndFormat(String diagnosisSetIds) {
```
Comment (mks-d) [inline api/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesByObsElement.java]: ??

File: api-pre2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java (hunk 8 above)
Comment (Ruhanga) [reply api-pre2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java]: Yes, of the three `EncounterDiagnoses*` on is for the `EncounterDiagnosesByObsElement`, and then for the other two `EncounterDiagnosesElement`s, one of them is used by versions of core less than 2.2 .

File: api/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesByObsElement.java (hunk 9 above)
Comment (Ruhanga) [reply api/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesByObsElement.java]: This is here for the `EncounterDiagnosesByObsElement` as explained above.

**Comment by Ruhanga** [comment]
@mks-d, is this fine to merge?

**Comment by Ruhanga** [comment]
The variable and method (`fragmentConfig & validateAndFormat()`) involved are internally used by the `EncounterDiagnoses*` elements and that an integration test is quite easier to test with. Second is that this seemed a good way to test changes on both the element and tag handler together.
